chain/chain/src/flat_storage_resharder.rs

```diff
@@ -460,6 +476,180 @@ impl FlatStorageResharder {
 
         Ok(iter)
     }
+
+    /// Task to perform catchup and creation of a flat storage shard spawned from a previous
+    /// resharding operation. This may be a long operation time-wise.
+    pub fn shard_catchup_task(
+        &self,
+        shard_uid: ShardUId,
+        flat_head_block_hash: CryptoHash,
+        chain_store: &ChainStore,
+    ) -> FlatStorageReshardingTaskStatus {
+        info!(target: "resharding", ?shard_uid, ?flat_head_block_hash, "flat storage shard catchup task started");
+        // If flat storage current status is beyond the chain final block, try again later.
+        let chain_final_head = chain_store.final_head().unwrap();
+        let height_result = chain_store.get_block_height(&flat_head_block_hash);
+        let Ok(height) = height_result else {
+            error!(target: "resharding", ?shard_uid, ?flat_head_block_hash, err = ?height_result.unwrap_err(), "can't find flat head height!");
+            return FlatStorageReshardingTaskStatus::Failed;
+        };
+        if height > chain_final_head.height {
+            info!(target = "resharding", ?height, chain_final_height = ?chain_final_head.height, "flat head beyond chain final tip: postponing flat storage shard catchup task");
+            self.scheduler.send(ReshardingRequest::FlatStorageShardCatchup {
```

Comment:
Wait, if this check is saying height is already beyond final head, why are we calling the schedule here?

Reply:
Oh this is a surprise I got from test loop resharding test.
It shouldn't happen in practice but the logic is the following: 
If the split operation is very fast, since we split parent flat head and deltas together, the children flat storage after the split may have their flat head beyond the final block height. We don't want that because of the invariant that flat head <= final block height, so we delay the catchup until the canonical chain contains the children flat head hash.

Reply:
Hmm, I see, well, I don't think this should be the solution. Instead what we should do is delay and retry the original split flat storage request (the run_later logic) for FlatStorageSplitShard and then we can hopefully remove this code and the retry logic for `FlatStorageShardCatchup` (assuming FlatStorageShardCatchup is called AFTER split is completed)